### PR TITLE
Dev/azure oauth

### PR DESCRIPTION
**Pull request by peterychang**
Add capability to use OAUTH authorization for eventhub and blob storage access

### CODE REVIEW

File: cmake/Modules/AzureVcpkg.cmake
Comment (jackgerrits) [inline cmake/Modules/AzureVcpkg.cmake:1]: rlclientlib already consumes dependencies from vcpkg. Why is this file needed?

File: examples/rl_sim_cpp/azure_credentials.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#include <azure/identity/azure_cli_credential.hpp>
+
+#include <chrono>
+#include <string>
+
+class AzureCredentials
+{
+public:
+  AzureCredentials();
+  int get_credentials(std::string& token_out, std::chrono::system_clock::time_point& expiry_out,
```
Comment (jackgerrits) [inline examples/rl_sim_cpp/azure_credentials.h:19]: Out params should be last

File: vcpkg.json [hunk 2]
```diff
@@ -21,7 +22,17 @@
     "openssl",
     "rapidjson",
     "spdlog",
-    "zlib"
+    "zlib",
+    "azure-storage-blobs-cpp",
+    "azure-identity-cpp"
```
Comment (jackgerrits) [inline vcpkg.json:27]: These aren't needed right?

File: cmake/Modules/AzureVcpkg.cmake
Comment (peterychang) [reply cmake/Modules/AzureVcpkg.cmake:1]: Builds were failing to find the module whenever I tried building without it. I'll take a closer look once I have a nuget out for Chad to test against

File: examples/rl_sim_cpp/azure_credentials.h (hunk 1 above)
Comment (peterychang) [reply examples/rl_sim_cpp/azure_credentials.h:19]: Thats our standard here? I'll get that fixed

File: vcpkg.json (hunk 2 above)
Comment (peterychang) [reply vcpkg.json:27]: azure-identity is needed for the example code, but the blobs aren't. How should we handle that?

Comment (jackgerrits) [reply vcpkg.json:27]: I'm not sure. Vcpkg must have some way to override this on the command line or something. It nothing else we can use manifest features like we do in vw. But this shouldn't make it into the default set.

File: examples/rl_sim_cpp/azure_credentials.h (hunk 1 above)
Comment (jackgerrits) [reply examples/rl_sim_cpp/azure_credentials.h]: I thought it was a pretty common standard, not sure if we have a convention or not in this codebase

File: vcpkg.json (hunk 2 above)
Comment (peterychang) [reply vcpkg.json:27]: I'll look into ways to remove conditionally remove it in the manifest

File: cmake/Modules/AzureVcpkg.cmake
Comment (peterychang) [reply cmake/Modules/AzureVcpkg.cmake:1]: Figured it out, misread one of the docs, you either use vcpkg OR use this. Removed superfluous stuff

**Review by jackgerrits — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 3]
```diff
@@ -33,6 +33,10 @@ if(RL_BUILD_BENCHMARKS)
   list(APPEND VCPKG_MANIFEST_FEATURES "benchmarks")
 endif()
 
+option(RL_LINK_AZURE_LIBS "Whether to build components requiring the use of Azure libraries. Requires C++14 or greater" OFF)
```
[the remark this answers is not in the document]
Comment (peterychang) [reply CMakeLists.txt:36]: I was planning on keeping it at least kind of generic in case there was ever a need to extend it with other azure libs

File: examples/rl_sim_cpp/azure_credentials.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+#  include "configuration.h"
+#  include "future_compat.h"
+
+#  include <azure/identity/azure_cli_credential.hpp>
+#  include <azure/identity/default_azure_credential.hpp>
+#  include <chrono>
+#  include <memory>
+#  include <string>
+
+class AzureCredentials
```
[the remark this answers is not in the document]
Comment (peterychang) [reply examples/rl_sim_cpp/azure_credentials.h:14]: theoretically it can support any azure credential. There's just a bug with the DefaultCredentials that I couldn't use it for the CLI stuff here

**Review by jackgerrits — APPROVED** [review]
